hummingbot/data_feed/candles_feed/candles_base.py

```diff
@@ -161,29 +162,34 @@ def load_candles_from_csv(self, data_path: str):
         self._candles.extendleft(df.values.tolist())
 
     async def get_historical_candles(self, config: HistoricalCandlesConfig):
+        candles_df = pd.DataFrame()
         try:
             await self.initialize_exchange_data()
-            all_candles = []
-            current_end_time = config.end_time + self.interval_in_seconds
-            current_start_time = config.start_time - self.interval_in_seconds
+            current_end_time = self._round_timestamp_to_interval_multiple(config.end_time)
+            current_start_time = self._round_timestamp_to_interval_multiple(config.start_time)
             while current_end_time >= current_start_time:
                 missing_records = int((current_end_time - current_start_time) / self.interval_in_seconds)
-                fetched_candles = await self.fetch_candles(end_time=current_end_time, limit=missing_records)
-                if fetched_candles.size <= 1:
+                candles = await self.fetch_candles(start_time=current_start_time,
+                                                   end_time=current_end_time,
+                                                   limit=missing_records)
+                if candles.size <= 1 or missing_records == 0:
                     break
-                all_candles.append(fetched_candles)
-                last_timestamp = self.ensure_timestamp_in_seconds(
-                    fetched_candles[0][0])  # Assuming the first column is the timestamp
-                current_end_time = last_timestamp - self.interval_in_seconds
-                self.check_candles_sorted_and_equidistant(all_candles)
-            final_candles = np.concatenate(all_candles[::-1], axis=0) if all_candles else np.array([])
-            candles_df = pd.DataFrame(final_candles, columns=self.columns)
-            candles_df.drop_duplicates(subset=["timestamp"], inplace=True)
+                candles = candles[candles[:, 0] <= current_end_time]
+                current_end_time = self.ensure_timestamp_in_seconds(candles[0][0])
+                fetched_candles_df = pd.DataFrame(candles, columns=self.columns)
+                candles_df = pd.concat([fetched_candles_df, candles_df])
+                candles_df.drop_duplicates(subset=["timestamp"], inplace=True)
+                candles_df.reset_index(drop=True, inplace=True)
+                self.check_candles_sorted_and_equidistant(candles_df.values)
             candles_df = candles_df[
                 (candles_df["timestamp"] <= config.end_time) & (candles_df["timestamp"] >= config.start_time)]
             return candles_df
+        except ValueError as e:
+            self.logger().error(f"Error fetching historical candles: {str(e)}")
+            raise e
         except Exception as e:
             self.logger().exception(f"Error fetching historical candles: {str(e)}")
+            raise e
 
     def check_candles_sorted_and_equidistant(self, candles: np.ndarray):
```

Comment:
hmm i'm not getting that error on live performance

![image](https://github.com/user-attachments/assets/2580100e-bb8b-4439-bf7a-5ceb56d5cc94)

can you please provide some input to reproduce it?

Reply:
It maybe related to the existing bug. It is not systematic, or at least initially I thought it was a Kraken issue. I printed the candle array whenever I get a reset and saw that only a few steps at the end had larger steps, the fact that is was often several steps seemed to indicate that it was not just the `first`/`last` issue

After I made a local patch (similar to yours, but more clunky) I no longer get it on my 15+ pairs test

Reply:
one possible factor can be the absence of trades over an entire interval in seconds, since candles update via websocket according to market activity. in low liquidity markets this represents an issue, that will be handled soon when i figure out how to handle it through [import this](https://calmcode.io/til/python-import-this) approach